src/vos/vos_tree.c

```diff
@@ -786,6 +739,98 @@ static struct vos_btr_attr vos_btr_attrs[] = {
 	},
 };
 
+static int
+tree_open_create(struct vos_object *obj, enum vos_tree_class tclass, int flags,
+		 struct vos_krec_df *krec, daos_handle_t *sub_toh)
+{
+	struct umem_attr	*uma = vos_obj2uma(obj);
+	struct vea_space_info	*info = obj->obj_cont->vc_pool->vp_vea_info;
+	daos_handle_t		 coh = vos_cont2hdl(obj->obj_cont);
+	int			 expected_flag;
+	int			 unexpected_flag;
+	int			 rc = 0;
+
+	if (flags & SUBTR_EVT) {
+		expected_flag = KREC_BF_EVT;
+		unexpected_flag = KREC_BF_BTR;
+	} else {
+		expected_flag = KREC_BF_BTR;
+		unexpected_flag = KREC_BF_EVT;
+	}
+
+	if (krec->kr_bmap & unexpected_flag) {
+		if (flags & SUBTR_CREATE) {
+			D_ERROR("Mixing single value and array not allowed\n");
+			rc = -DER_NO_PERM;
+			goto out;
+		}
+		D_DEBUG(DB_TRACE, "Attempt to fetch wrong value type\n");
+		rc = -DER_NONEXIST;
+		goto out;
+	}
+
+	if (krec->kr_bmap & expected_flag) {
+		if (flags & SUBTR_EVT) {
+			rc = evt_open_inplace(&krec->kr_evt, uma, coh, info,
+					      sub_toh);
+		} else {
+			rc = dbtree_open_inplace_ex(&krec->kr_btr, uma, coh,
+						    info, sub_toh);
+		}
+		if (rc != 0)
+			D_ERROR("Failed to open tree: %d\n", rc);
+
+		goto out;
+	}
+
+	D_ASSERT(flags & SUBTR_CREATE);
+
+	if (flags & SUBTR_EVT) {
+		rc = evt_create_inplace(EVT_FEAT_DEFAULT, VOS_EVT_ORDER,
+					uma, &krec->kr_evt, coh, sub_toh);
+		if (rc != 0) {
+			D_ERROR("Failed to create evtree: %d\n", rc);
+			goto out;
+		}
+	} else {
+		struct vos_btr_attr	*ta;
+		uint64_t		 tree_feats = 0;
+
+		/* Step-1: find the btree attributes and create btree */
+		if (tclass == VOS_BTR_DKEY) {
+			uint64_t	obj_feats;
+
+			/* Check and setup the akey key compare bits */
+			obj_feats = daos_obj_id2feat(obj->obj_df->vo_id.id_pub);
+			tree_feats = (uint64_t)obj_feats << VOS_OFEAT_SHIFT;
+			if (obj_feats & DAOS_OF_AKEY_UINT64)
+				tree_feats |= VOS_KEY_CMP_UINT64_SET;
+			else if (obj_feats & DAOS_OF_AKEY_LEXICAL)
+				tree_feats |= VOS_KEY_CMP_LEXICAL_SET;
+		}
+
+
+		ta = obj_tree_find_attr(tclass);
+
+		D_DEBUG(DB_TRACE, "Create dbtree %s feats 0x"DF_X64"\n",
+			ta->ta_name, tree_feats);
+
+		rc = dbtree_create_inplace_ex(ta->ta_class, tree_feats,
+					      ta->ta_order, uma, &krec->kr_btr,
+					      coh, sub_toh);
+		if (rc != 0) {
+			D_ERROR("Failed to create btree: %d\n", rc);
+			goto out;
+		}
+	}
+
+	/* NB: Should probably be transactional */
```

Comment:
Is the tx_add_range() called upon vos_krec_df or not? I think kr_bmap updating should be in transaction for sure.

Reply:
In the case of create, it will be in the transaction because the record will have been allocated.   I suppose you could argue that it's not in the same transaction because there is a nest around the btree code but the answer is yes, it is added by way of allocation.

Reply:
I sensed kind of uncertainty from the comment ('should probably'), I think we'd just remove the comment or change it from a 'question' to a definite description.

Reply:
Reworded the comment